### PR TITLE
Respond all submission records in an exam's transcript

### PR DESCRIPTION
**Pull request by Johnny850807**
Respond all submission records in an exam's transcript

### CODE REVIEW

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/MapUtils.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.Map;
+
+/**
+ * @author - johnny850807@gmail.com (Waterball)
+ */
+public class MapUtils {
+
+    public static <K, V> Map<K, V> zip(K[] keys, V[] values) {
+        Map<K, V> map = new HashMap<>();
+        for (int i = 0; i < keys.length; i++) {
```
Comment (Wally5077) [inline Commons/src/main/java/tw/waterball/judgegirl/commons/utils/MapUtils.java:15]: NIT : keys length might less than values length

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java [hunk 2]
```diff
@@ -114,6 +109,27 @@
...
+                                              BiFunction<L, R, T> zipAndMap) {
+        return range(0, left.length)
+                .mapToObj(i -> zipAndMap.apply(left[i], right[i]))
+                .collect(toList());
+    }
+
+    public static <T, L, R> List<T> zipToList(Collection<L> left, Collection<R> right,
+                                              BiPredicate<L, R> matching,
+                                              BiFunction<L, R, T> zipAndMap) {
+        List<T> list = new ArrayList<>();
```
Comment (Wally5077) [inline Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java:122]: var list = new LinkedList<>();

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java [hunk 3]
```diff
@@ -22,20 +25,34 @@
...
+    public CalculateExamScoreUseCase(ExamRepository examRepository, StudentServiceDriver studentServiceDriver, SubmissionServiceDriver submissionServiceDriver) {
         super(examRepository);
         this.studentServiceDriver = studentServiceDriver;
+        this.submissionServiceDriver = submissionServiceDriver;
     }
 
     public void execute(int examId, Presenter presenter) {
         Exam exam = findExam(examId);
-        List<Record> records = examRepository.findAllRecordsInAnExam(examId);
+        List<ExamineeRecord> examineeRecords = getExamineeRecords(exam);
```
Comment (Wally5077) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java:38]: List<ExamineeRecord> -> var

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java [hunk 4]
```diff
@@ -22,20 +25,34 @@
...
-        List<ExamineeRecord> examineeRecords = examineeRecords(exam, records, examinees);
-        double averageScore = average(examineeRecords, ExamineeRecord::getTotalScore);
-        present(presenter, exam, examineeRecords, averageScore);
+        return examineeRecords(examinees, questionRecords);
+    }
+
+    private List<QuestionRecord> findQuestionRecords(Exam exam) {
+        String[] submissionIds = mapToList(examRepository.findAllRecordsInAnExam(exam.getId()),
+                Record::getSubmissionId).toArray(new String[0]);
+        List<Submission> submissions = mapToList(submissionServiceDriver.getSubmissions(submissionIds), SubmissionView::toEntity);
```
Comment (Wally5077) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java:52]: String[] submissionIds -> var submissionIds
List<Submission> submissions -> var submissions

File: Domain/Submission/src/main/java/tw/waterball/judgegirl/submission/domain/usecases/GetSubmissionsUseCase.java [hunk 5]
```diff
@@ -33,10 +33,14 @@ public GetSubmissionsUseCase(SubmissionRepository submissionRepository) {
 
     public void execute(SubmissionQueryParams queryParams, Presenter presenter) {
         List<Submission> submissions = submissionRepository.query(queryParams);
```
Comment (Wally5077) [inline Domain/Submission/src/main/java/tw/waterball/judgegirl/submission/domain/usecases/GetSubmissionsUseCase.java:35]: List<Submission> submissions -> var submissions

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptPresenter.java [hunk 6]
```diff
@@ -16,20 +23,35 @@ public void showExam(Exam exam) {
     }
 
     @Override
-    public void addRecord(ExamineeRecord examineeRecord) {
-        builder.examineeRecord(examineeRecord.getExaminee().getEmail(),
-                new TranscriptView.ExamineeRecord(
-                        examineeRecord.getTotalScore(), examineeRecord.getScores()
-                ));
+    public void showRecords(List<ExamineeRecord> examineeRecords) {
+        Map<String, TranscriptView.ExamineeRecordView> stringExamineeRecordMap = mapEmailToExamineeRecord(examineeRecords);
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptPresenter.java:27]: Map<String, TranscriptView.ExamineeRecordView> stringExamineeRecordMap -> var emailExamineeRecordMap

File: Spring-Boot/Spring-Boot-Submission/src/main/java/tw/waterball/judgegirl/springboot/submission/presenters/SubmissionsPresenter.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * @author - johnny850807@gmail.com (Waterball)
+ */
+public class SubmissionsPresenter implements GetSubmissionsUseCase.Presenter {
+    private List<SubmissionView> submissionViews;
+
+    @Override
+    public void showSubmissions(List<Submission> submissions) {
+        this.submissionViews = submissions.stream()
+                .map(SubmissionView::toViewModel).collect(Collectors.toList());
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Submission/src/main/java/tw/waterball/judgegirl/springboot/submission/presenters/SubmissionsPresenter.java:19]: mapToList

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: API/Submission-API/src/main/java/tw/waterball/judgegirl/submissionapi/clients/FakeSubmissionServiceDriver.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    public List<SubmissionView> getSubmissions(int problemId, int studentId, Map<String, String> bagQueryParameters) {
+        return submissions.values().stream()
+                .filter(submission -> submission.getProblemId() == problemId &&
+                        submission.getStudentId() == studentId && filter(submission, bagQueryParameters))
+                .map(SubmissionView::toViewModel)
+                .collect(toList());
+    }
+
+    private boolean filter(Submission submission, Map<String, String> bagQueryParameters) {
+        Bag bag = submission.getBag();
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply API/Submission-API/src/main/java/tw/waterball/judgegirl/submissionapi/clients/FakeSubmissionServiceDriver.java:60]: I think this is okay, as Bag is a short word.

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/MapUtils.java (hunk 1 above)
Comment (Johnny850807) [reply Commons/src/main/java/tw/waterball/judgegirl/commons/utils/MapUtils.java:15]: that's okay, as if are mapping the keys to values, the values that have no keys map to will be ignored.

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java (hunk 2 above)
Comment (Johnny850807) [reply Commons/src/main/java/tw/waterball/judgegirl/commons/utils/StreamUtils.java:122]: ArrayList will be more efficient as we have known its size

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/MapUtils.java (hunk 1 above)
Comment (Wally5077) [reply Commons/src/main/java/tw/waterball/judgegirl/commons/utils/MapUtils.java:15]: but it would throw IndexOutOfBoundException

**Review by Wally5077 — APPROVED** [review]
LGTM

**Pull request merged**